File: app/src/test/java/com/hana053/micropost/ui/pages/micropostnew/MicropostNewServiceTest.java
package com.hana053.micropost.ui.pages.micropostnew;

import com.hana053.micropost.domain.Micropost;
import com.hana053.micropost.interactors.MicropostInteractor;
import com.hana053.micropost.testing.RobolectricBaseTest;

import org.junit.Test;

import rx.Observable;
import rx.observers.TestSubscriber;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class MicropostNewServiceTest extends RobolectricBaseTest {

    private final MicropostInteractor micropostInteractor = mock(MicropostInteractor.class);
    private final MicropostNewService micropostNewService = new MicropostNewServiceImpl(micropostInteractor);

    @Test
    public void shouldCreatePost() {
        Observable<Micropost> response = Observable.just(new Micropost(1, "", 0, null));
        when(micropostInteractor.create(any(Micropost.class))).thenReturn(response);

        TestSubscriber<Micropost> testSubscriber = new TestSubscriber<>();
        micropostNewService.createPost("test content").subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
    }
}

File: app/src/test/java/com/hana053/micropost/pages/micropostnew/MicropostNewServiceTest.kt
package com.hana053.micropost.pages.micropostnew

import com.hana053.micropost.domain.Micropost
import com.hana053.micropost.interactors.MicropostInteractor
import com.hana053.micropost.services.HttpErrorHandler
import com.hana053.micropost.testing.RobolectricBaseTest
import com.hana053.micropost.testing.TestMicropost
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import org.junit.Test
import org.mockito.Mockito.`when`
import rx.Observable


class MicropostNewServiceTest : RobolectricBaseTest() {

    private val micropostInteractor = mock<MicropostInteractor>()
    private val httpErrorHandler = mock<HttpErrorHandler>()
    private val micropostNewService = MicropostNewService(
        micropostInteractor = micropostInteractor,
        httpErrorHandler = httpErrorHandler
    )

    @Test
    fun shouldCreatePost() {
        `when`(micropostInteractor.create(any()))
            .doReturn(Observable.just(TestMicropost))

        micropostNewService.createPost("my post").subscribe()
        advance()

        verify(micropostInteractor)
            .create(MicropostInteractor.MicropostRequest("my post"))
    }

    @Test
    fun shouldHandleErrorWhenCreatePost() {
        val error = Observable.error<Micropost>(RuntimeException())
        `when`(micropostInteractor.create(any())).doReturn(error)

        micropostNewService.createPost("my post").subscribe()
        advance()

        verify(httpErrorHandler).handleError(any<RuntimeException>())
    }
}